### PR TITLE
Feature: async_checkpoint

### LINKED ISSUE #578: async_checkpoint
Followup to #558 we want to provide an async checkpoint layer so we can eventually support jwt auth etc etc.

### PR DESCRIPTION
**Pull request by o0Ignition0o**
Fixes #578

This pr introduces `ServiceBuilder().async_checkpoint()` that takes an async callback that takes `S::Request` and outputs a `Result<Step<Request, Response>, BoxError>>`

`async_checkpoint` is the async version of `checkpoint`

This pr also adds an `AllowClientIdFromFileList` example, that reads asynchronously from a file, and matches an `x-client-id` header against the allowed clients list.

The purpose of this example was to use and test `async_checkpoint()` but one can easily derive usecases where we would use this mechanism to make asynchronous http calls for say authentication or something else.

### CODE REVIEW

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Error: Into<BoxError> + Send + Sync + 'static,
+{
+    type Service = AsyncCheckpointService<
+        BoxCloneService<Request, <S as Service<Request>>::Response, BoxError>,
+        Request,
+    >;
+
+    fn layer(&self, service: S) -> Self::Service {
+        let inner = Buffer::new(service, 20_000);
```
Comment (BrynCooke) [inline apollo-router-core/src/services/checkpoint/async_checkpoint.rs:69]: Suggest leaving buffer out for now and requiring a clone service.
We can then put the buffer call in ServiceBuilderExt and handle all the other similar layers in the same way.

**Review by BrynCooke — COMMENTED** [review]
Super cool! Let's pull out the buffer thing though. We can handle this centrally in ServiceBuilderExt

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+{
+    checkpoint_fn: Arc<
+        dyn Fn(
+                Request,
+            ) -> BoxFuture<
+                'static,
+                Result<Step<Request, <S as Service<Request>>::Response>, BoxError>,
+            > + Send
+            + Sync
+            + 'static,
```
Comment (bnjjj) [inline apollo-router-core/src/services/checkpoint/async_checkpoint.rs:25]: I'm missing trait alias A LOOOOOOT :(

**Review by bnjjj — APPROVED** [review]

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/services/checkpoint/async_checkpoint.rs:66]: fixed in 7b2b0159

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    type Service = AsyncCheckpointService<
+        BoxCloneService<Request, <S as Service<Request>>::Response, BoxError>,
+        Request,
+    >;
+
+    fn layer(&self, service: S) -> Self::Service {
+        let inner = Buffer::new(service, 20_000);
+        AsyncCheckpointService {
+            checkpoint_fn: Arc::clone(&self.checkpoint_fn),
+            inner: ServiceBuilder::new().service(inner).boxed_clone(),
```
Comment (BrynCooke) [inline apollo-router-core/src/services/checkpoint/async_checkpoint.rs:69]: Should just be able to store inner directly here.

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+#[derive(Clone)]
+#[allow(clippy::type_complexity)]
+pub struct AsyncCheckpointService<S, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Clone + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+{
+    inner: BoxCloneService<Request, <S as Service<Request>>::Response, BoxError>,
```
Comment (BrynCooke) [inline apollo-router-core/src/services/checkpoint/async_checkpoint.rs:83]: Inner is of type S

**Review by BrynCooke — CHANGES_REQUESTED** [review]
Still some stuff about buffering in the code.

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs (hunk 3 above)
Comment (BrynCooke) [reply apollo-router-core/src/services/checkpoint/async_checkpoint.rs:69]: Also need to remove the buffer.

Comment (o0Ignition0o) [reply apollo-router-core/src/services/checkpoint/async_checkpoint.rs:69]: nice catch it s a leftover indeed!

File: apollo-router-core/src/services/checkpoint/async_checkpoint.rs (hunk 4 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/services/checkpoint/async_checkpoint.rs]: fixed in bb7db24d

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**